arrayfire/array_api/array_object.py

```diff
@@ -0,0 +1,1093 @@
+        ndims = self.ndim
+
+        if isinstance(key, Array) and key == af_bool.c_api_value:
+            ndims = 1
+            if count(key) == 0:
+                return out
+
+        safe_call(backend.get().af_index_gen(
+            ctypes.pointer(out.arr), self.arr, c_dim_t(ndims), _get_indices(key).pointer))
+        return out
+
+    def __index__(self) -> int:
+        # TODO
+        return NotImplemented
+
+    def __int__(self) -> int:
+        # TODO
+        return NotImplemented
+
+    def __len__(self) -> int:
+        return self.shape[0] if self.shape else 0
+
+    def __setitem__(
+            self, key: Union[int, slice, Tuple[Union[int, slice, ], ...], Array],
+            value: Union[int, float, bool, Array], /) -> None:
+        # TODO
+        return NotImplemented  # type: ignore[return-value]  # FIXME
+
+    def __str__(self) -> str:
+        # TODO change the look of array str. E.g., like np.array
+        if not _in_display_dims_limit(self.shape):
+            return _metadata_string(self.dtype, self.shape)
+
+        return _metadata_string(self.dtype) + _array_as_str(self)
+
+    def __repr__(self) -> str:
+        # return _metadata_string(self.dtype, self.shape)
+        # TODO change the look of array representation. E.g., like np.array
+        return _array_as_str(self)
+
+    def to_device(self, device: Any, /, *, stream: Union[int, Any] = None) -> Array:
+        # TODO implementation and change device type from Any to Device
+        return NotImplemented
+
+    # Attributes
+
+    @property
+    def dtype(self) -> Dtype:
+        """
+        Data type of the array elements.
+
+        Returns
+        -------
+        out : Dtype
+            Array data type.
+        """
+        out = ctypes.c_int()
+        safe_call(backend.get().af_get_type(ctypes.pointer(out), self.arr))
+        return _c_api_value_to_dtype(out.value)
+
+    @property
+    def device(self) -> Any:
+        # TODO
+        return NotImplemented
+
+    @property
+    def mT(self) -> Array:
+        # TODO
+        return NotImplemented
+
+    @property
+    def T(self) -> Array:
+        """
+        Transpose of the array.
+
+        Returns
+        -------
+        out : Array
+            Two-dimensional array whose first and last dimensions (axes) are permuted in reverse order relative to
+            original array. The returned array must have the same data type as the original array.
+
+        Note
+        ----
+        - The array instance must be two-dimensional. If the array instance is not two-dimensional, an error
+        should be raised.
+        """
+        if self.ndim < 2:
+            raise TypeError(f"Array should be at least 2-dimensional. Got {self.ndim}-dimensional array")
+
+        # TODO add check if out.dtype == self.dtype
+        out = Array()
+        safe_call(backend.get().af_transpose(ctypes.pointer(out.arr), self.arr, False))
+        return out
+
+    @property
+    def size(self) -> int:
```

Comment:
Not following spec, should be Optional[Int]
Copy docstrings from https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.size.html#array_api.array.size

Reply:
https://mypy.readthedocs.io/en/stable/kinds_of_types.html#optional-types-and-the-none-type

Reply:
And still, we won't have a case where size equals `None`, but making the return `Optional` will lead us to major logical code changes if we want to keep typing work.
Mypy doesn't implement it in their source code as well https://github.com/numpy/numpy/blob/6f3e1f458e04d13bdd56cff5669f9fd96a25fb66/numpy/array_api/_array_object.py#L1088
My point is to keep `int` as an expected return and not try to imagine cases that likely never happen in the future with `None` return.
Or, let's switch to TDD and add valid test case for `None` return and then fix typing in the source code 🤷‍♂️